packages/dynamic_layouts/lib/src/staggered_layout.dart

```diff
@@ -0,0 +1,279 @@
+///    freely sized in the main and cross axis.
+///  * [DynamicSliverGridGeometry], which represents the size and position of a
+///    single tile in a grid.
+///  * [SliverGridDelegate.getLayout], which returns this object to describe the
+///    delegate's layout.
+///  * [RenderDynamicSliverGrid], which uses this class during its
+///    [RenderDynamicSliverGrid.performLayout] method.
+class SliverGridStaggeredTileLayout extends DynamicSliverGridLayout {
+  /// Creates a layyout with dynamic main axis extents and fixed cross axis
+  /// extents.
+  ///
+  /// All arguments must be not null. The [crossAxisCount] argument must be
+  /// greater than zero. [mainAxisSpacing], [crossAxisSpacing] and
+  /// [childCrossAxisExtent] must not be negative.
+  SliverGridStaggeredTileLayout({
+    required this.crossAxisCount,
+    required this.mainAxisSpacing,
+    required this.crossAxisSpacing,
+    required this.childCrossAxisExtent,
+    required this.scrollDirection,
+  })  : assert(crossAxisCount != null && crossAxisCount > 0),
+        assert(crossAxisSpacing != null && crossAxisSpacing >= 0),
+        assert(childCrossAxisExtent != null && childCrossAxisExtent >= 0) {
+    for (int i = 0; i < crossAxisCount; i++) {
+      _scrollOffsetForMainAxis[i] = 0.0;
+      _mainAxisCount[i] = 0;
+    }
+  }
+
+  /// The number of children in the cross axis.
+  final int crossAxisCount;
+
+  /// The number of logical pixels between each child along the main axis.
+  final double mainAxisSpacing;
+
+  /// The number of logical pixels between each child along the cross axis.
+  final double crossAxisSpacing;
+
+  /// The number of pixels from the leading edge of one tile to the trailing
+  /// edge of the same tile in the cross axis.
+  final double childCrossAxisExtent;
+
+  /// The axis along which the scroll view scrolls.
+  final Axis scrollDirection;
+
+  /// The collection of scroll offsets for every row or column across the main
+  /// axis. It includes the tiles' sizes and the spacing between them.
+  final Map<int, double> _scrollOffsetForMainAxis = <int, double>{};
+
+  /// The amount of tiles in every row or column across the main axis.
+  final Map<int, int> _mainAxisCount = <int, int>{};
+
+  /// Returns the row or column with the minimum extent for the next child to
+  /// be layed out.
+  int _getNextCrossAxisSlot() {
+    int nextCrossAxisSlot = 0;
+    double minScrollOffset = double.infinity;
+    _scrollOffsetForMainAxis.forEach((int mainAxisSeries, double scrollOffset) {
+      if (scrollOffset < minScrollOffset) {
+        nextCrossAxisSlot = mainAxisSeries;
+        minScrollOffset = scrollOffset;
+      }
+    });
+    return nextCrossAxisSlot;
+  }
+
+  /// Returns true if every child has crossed the [targetOffset] and false
+  /// if not.
+  @override
+  bool reachedTargetScrollOffset(double targetOffset) {
+    for (final double scrollOffset in _scrollOffsetForMainAxis.values) {
```

Comment:
I'm not sure if binary search could be used because the scrollOffsets are not in order, but let me check if there are more efficient approaches for this

Reply:
This sounds like a good follow up item, to see if we can optimize. Feel free to file a bug with the `p: dynamic_layouts` label in flutter/flutter so we can track it. I know you have a few follow up changes already planned for after this PR, so we can iterate after you land your initial version. :)